doc/running/caching.rst

```diff
@@ -0,0 +1,52 @@
+Caching in mirgecom
+===================
+
+
+OpenCL Kernels
+--------------
+
+OpenCL kernels are cached on hard disks on multiple levels in a mirgecom
+execution. This has the advantage of reducing the compilation time of kernels
+when running the same driver multiple times, but can lead to inconsistent
+performance results, especially when comparing multiple executions against
+each other, as subsequent executions can appear faster than earlier ones due
+to caching effects.
+
+
+The following list discusses mirgecom-related packages that use caching.
+
+Loopy
++++++
+
+Loopy stores generated PyOpenCL kernels in ``$XDG_CACHE_HOME`` by default. You can
+export ``LOOPY_NO_CACHE=1`` to disable caching. See
+`here <https://github.com/inducer/loopy/blob/e21e8f85d289abbca27ac6abfd71874155fa49da/loopy/__init__.py#L402-L406>`__
+for details.
+
+PyOpenCL
+++++++++
+
+PyOpenCL stores generated OpenCL kernels in ``$XDG_CACHE_HOME`` by default. You can
+export ``PYOPENCL_NO_CACHE=1`` to disable caching. See
+`here <https://documen.tician.de/pyopencl/runtime_program.html#envvar-PYOPENCL_NO_CACHE>`__
+for details.
+
+POCL
+++++
+
+POCL stores compilation results in ``$POCL_CACHE_DIR`` or ``$XDG_CACHE_HOME``
+by default. You can export ``POCL_KERNEL_CACHE=0`` to disable caching. See
+`here <http://portablecl.org/docs/html/env_variables.html>`__ for details.
```

Comment:
POCL has its own "from-binary" cache and its own "from-source" cache. So, in a sense, PyOpenCL's cache is only making things worse, by forcing a miss on second run via the use of the (still-cold) from-binary cache. We should probably address that, probably by turning off caching in PyOpenCL. (I dimly recall adding code that does that, but I can't find it now.)

Reply:
I don't see where POCL stores source code.

Reply:
Pocl doesn't cache source code. OpenCL has `clCreateProgramWithSource` and `clCreateProgramWithBinary` (actually also `clCreateProgramWithIL`). PyOpenCL uses the first the first time around and caches the (OpenCL) binary it retrieves. The second time the same source is compiled, it hands pocl the (CL) binary it retrieved. IIUC, both paths have separate caches in pocl.

- Please explain this in the doc.
- I sometimes get the feeling that just feeding pocl the same source again instead of caching the binary may lead to faster second-round compile times. (It may be what's behind our "only the third run is fast" problem.) Please investigate, and, if applicable prepare a patch to PyOpenCL that skips CL binary caching for pocl if applicable.

Reply:
I added a note regarding this in d798d7b9. As far as I can tell, both cases are cached in the same top-level dir in pocl.